Travis: report coverage with all builds

blueyed opened the pull request:
https://codecov.io/gh/pytest-dev/pytest/tree/coverage

TODO:

- [x] https://github.com/pytest-dev/pytest/pull/3877 should be merged first, then this one rebased.
- [x] bring back AppVeyor

blueyed commented:
It seems to be problematic with py27/py26 (https://travis-ci.org/pytest-dev/pytest/builds/280216101).
I've triggered a full build now again.

Is this something we want to have in general?

nicoddemus commented on .travis.yml:
```diff
@@ -5,11 +5,15 @@ python:
 install:
   - pip install --upgrade --pre tox
+  - |
+    pip install tox
```
tox is already installed in the line above, so this is not really needed

nicoddemus commented on .travis.yml:
```diff
@@ -5,11 +5,15 @@ python:
   - pip install --upgrade --pre tox
+  - |
+    pip install tox
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
```
Pardon my lack of bash-fu, what does this expression inside `if` does?

blueyed replied:
TOXENV starts with "py".

nicoddemus replied:
Thanks, and what does `"${TOXENV#py}"` do? (I tried googling it but it wasn't entirely obvious what to search for).

blueyed replied:
It substitutes "py" at the beginning.
Yes, shell programming can be ugly.  See "Parameter Expansion" in bash's man page.

nicoddemus replied:
Thanks!

nicoddemus commented on tox.ini:
```diff
@@ -64,57 +67,65 @@ deps =
+    {env:TOX_EXTRA_DEPS:}
 commands =
-    pytest -ra test_pdb.py test_terminal.py test_unittest.py
+    {env:TOX_COVERAGE_RUN:} pytest -ra test_pdb.py test_terminal.py test_unittest.py
```
Looking at the Travis build, `py27/py36-trial` seem to be missing this configuration

nicoddemus commented on .coveragerc:
```diff
@@ -1,4 +1,5 @@
 [run]
+source = _pytest
 omit = 
     # standlonetemplate is read dynamically and tested by test_genscript
```
Oh this does not exist anymore and can be removed. 😁

blueyed replied:
standlonetemplate?

Apart from that I think we should also include `testing` (the tests themselves).  Anything else?

nicoddemus replied:
"standlonetemplate" was used by the removed `genscript` feature.

> Apart from that I think we should also include testing (the tests themselves). Anything else?

Not sure, then `xfail` tests which are never meant to run because they reproduce issues not yet fixed will drop the coverage. In my head "coverage" does not include the tests. What's your opinion?

blueyed replied:
I think unexpected missing coverage is a good sign of something being wrong.
Those xfail tests should be marked with "pragma: no cover" then.

nicoddemus commented on .travis.yml:
```diff
@@ -5,11 +5,15 @@ python:
+  - |
+    pip install tox
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
+      export TOX_COVERAGE_RUN="coverage run -m"
```
At first glance I thought `TOX_EXTRA_DEPS` was a feature of tox itself. Perhaps we should add a PYTEST_ prefix to avoid confusion? While at it, perhaps there is a way to add extra dependencies using some tox feature?

nicoddemus replied:
Just remembered that we used `_PYTEST_SETUP_SKIP_PLUGGY_DEP` in [setup.py](https://github.com/pytest-dev/pytest/blob/features/setup.py#L49) for this type of "setup/buid only flag used internally".

`_PYTEST_TOX_COVERAGE_RUN`?

blueyed replied:
Makes sense.  Will rename it to `_PYTEST_TOX_COVERAGE_RUN` and `_PYTEST_TOX_EXTRA_DEPS` then?

nicoddemus replied:
Sounds good, thanks

blueyed commented:
Failures on py2, pypy and py33 are all:
```
=========================== short test summary info ============================
FAIL testing/code/test_excinfo.py::TestTraceback_f_g_h::()::test_traceback_cut_excludepath
FAIL testing/python/collect.py::TestModule::()::test_show_traceback_import_error[0]
FAIL testing/python/collect.py::TestModule::()::test_show_traceback_import_error[1]
FAIL testing/python/collect.py::TestTracebackCutting::()::test_traceback_argsetup
```
e.g. https://travis-ci.org/pytest-dev/pytest/jobs/280479131#L599

Any idea?

blueyed commented:
@nicoddemus 
Can you enable the codecov bot user on https://codecov.io/gh/pytest-dev/pytest?

nicoddemus commented:
> Can you enable the codecov bot user on https://codecov.io/gh/pytest-dev/pytest?

Done! If we will use `codecov` we should disable `coveralls` (after this is merged)

blueyed commented:
The problem with `test_traceback_cut_excludepath` is that the traceback entry looks different when run with `-m`, also when using `python -m pytest` (and we use `coverage run -m`).

`<TracebackEntry _pytest/python_api.py:580>` vs `<TracebackEntry /home/user/Vcs/pytest/.tox/py27/lib/python2.7/site-packages/_pytest/python_api.py:580>`.
Is that something that should get fixed for `excinfo.traceback`?
Looks like it uncovered an actual issue?!

blueyed commented:
This works around it:

```diff
diff --git i/testing/code/test_excinfo.py w/testing/code/test_excinfo.py
index f8f8a036..cc512a2d 100644
--- i/testing/code/test_excinfo.py
+++ w/testing/code/test_excinfo.py
@@ -162,9 +162,12 @@ def test_traceback_cut(self):
         assert len(newtraceback) == 1
 
     def test_traceback_cut_excludepath(self, testdir):
+        ptf = py.path.local(pytest.__file__)
+        if not ptf.exists():
+            pytest.skip('pytest.__file__ does not point to an actual file.')
+        basedir = ptf.dirpath()
         p = testdir.makepyfile("def f(): raise ValueError")
         excinfo = pytest.raises(ValueError, "p.pyimport().f()")
-        basedir = py.path.local(pytest.__file__).dirpath()
         newtraceback = excinfo.traceback.cut(excludepath=basedir)
         for x in newtraceback:
             if hasattr(x, 'path'):
```

Looks like an issue with Python 2 where `pytest.__file__` is not correct in this case.

nicoddemus commented:
I gotta say I like codecov's report better. 😄

nicoddemus commented on .travis.yml:
```diff
@@ -49,6 +52,18 @@ matrix:
+      pip install codecov coveralls
+      coverage xml
+      coverage report -m
+      coveralls
```
Should we drop `coveralls` if we are migrating to `codecov`?

blueyed replied:
I think it is good to keep it for comparison, at least for a while.

nicoddemus commented on tox.ini:
```diff
@@ -19,29 +19,32 @@ envlist =
+    {env:TOX_EXTRA_DEPS:}
 commands =
-    pytest -n3 -ra --runpytest=subprocess {posargs:testing}
+    {env:TOX_COVERAGE_RUN:} pytest -n3 -ra --runpytest=subprocess {posargs:testing}
```
I'm seeing this warning from tox:

```
WARNING:test command found but not installed in testenv
  cmd: /usr/bin/env
  env: /home/travis/build/pytest-dev/pytest/.tox/py27
```

Do we have to whitelist `env`?

The-Compiler replied:
Couldn't we just use [pytest-cov](https://github.com/pytest-dev/pytest-cov) instead?

RonnyPfannschmidt replied:
pytest-cov does incomplete reports due to late plugin startup, its a longstanding pytest issue

nicoddemus commented on tox.ini:
```diff
@@ -183,23 +196,13 @@ commands =
-    coverage run --source=_pytest -m pytest testing
-    coverage report -m
-    coveralls
+    {env:TOX_COVERAGE_RUN:} {envpython} create_executable.py
```
We can drop coverage from this env I believe

RonnyPfannschmidt commented on tox.ini:
```diff
@@ -19,29 +19,32 @@ envlist =
 
 [testenv]
-commands = pytest --lsof -ra {posargs:testing}
+commands = {env:TOX_COVERAGE_RUN:} pytest --lsof -ra {posargs:testing}
```
can we just introduce a coverage factor for this?

blueyed replied:
@RonnyPfannschmidt 
There is no way to have a command by default and then another for a factor, is there? (where the command comes not from the default testenv).

blueyed replied:
Will do it via `setenv`:
```
[testenv]
commands = {env:_TOX_COVERAGE_RUN:} pytest --lsof -ra {posargs:testing}
    coverage: coverage report -m --skip-covered
setenv =
    coverage: _TOX_COVERAGE_RUN=env COVERAGE_FILE={toxinidir}/.coverage coverage run -a -m
whitelist_externals =
    coverage: env
```

RonnyPfannschmidt commented on .travis.yml:
```diff
@@ -5,11 +5,14 @@ python:
   - pip install --upgrade --pre tox
+  - |
+    if [[ "${TOXENV#py}" != "$TOXENV" ]]; then
+      export TOX_COVERAGE_RUN="env COVERAGE_FILE=$PWD/.coverage coverage run -a -m"
```
at frist glance this "breaks" at least for the xdist envs

RonnyPfannschmidt commented on tox.ini:
```diff
@@ -19,29 +19,32 @@ envlist =
     nose
     mock
     requests
+    {env:TOX_EXTRA_DEPS:}
```
dito

blueyed commented:
@nicoddemus 
See https://github.com/pytest-dev/pytest/pull/2800#issuecomment-332635907.

RonnyPfannschmidt commented:
instead of falling out for pytest in some cases

lets use `_pytest` instead - its not being used as module

nicoddemus commented:
> lets use _pytest instead - its not being used as module

You mean in the production code or the test?

RonnyPfannschmidt commented:
in `testing/code/test_excinfo.py`

nicoddemus commented:
@blueyed I think we have fixed the errors you were experiencing, would you like to give this another shot?

blueyed commented:
@nicoddemus 
The problem(s) with py27 seems to be still there: https://travis-ci.org/pytest-dev/pytest/jobs/314001060#L621.
But I had rebased it against master - now for features.

nicoddemus commented:
@blueyed gentle ping if you still want to give this a shot; `coveralls` is acting strange this days so `codecov` would be a welcome change. 😁

blueyed commented:
py27-xdist fails (https://travis-ci.org/pytest-dev/pytest/jobs/420593988#L668):

> Couldn't parse '/home/travis/build/pytest-dev/pytest/.tox/py27-xdist/lib/python2.7/site-packages/nose/config.py' as Python source: 'invalid syntax' at line 68

py36-freeze fails (https://travis-ci.org/pytest-dev/pytest/jobs/420594002#L665):

> No data to report.

We should skip pypy, because it is very slow with coverage reporting (https://travis-ci.org/pytest-dev/pytest/jobs/420594000) - takes > 20mins for 68% already.

asottile commented on sitecustomize.py:
```diff
@@ -0,0 +1,9 @@
+"""
+import coverage
+
+coverage.process_startup()
```
I've used `coverage-enable-subprocess` with success before -- it essentially does this but with a `.pth` file (no need for a relative `sitecustomize.py` file)

blueyed replied:
Awesome, will use that.

asottile commented on .travis.yml:
```diff
@@ -62,15 +59,44 @@ jobs:
+    if [[ "$PYTEST_NO_COVERAGE" != 1 ]]; then
+      set -ex
+      # Massage .coveragerc to use source from tox env (only).
+      sed -i "s~source~source = $(ls $PWD/.tox/$TOXENV/lib/python*/site-packages/_pytest/),$PWD/testing" .coveragerc
```
one trick you can use here is `.coveragerc` supports environment variables.

I'm trying to find a not-closed-source example for you but this is the rough sketch:

```ini
setenv =
    TOP={toxinidir}
```

and then in `.coveragerc`

```
source = $TOP
```

(I think `$PWD` probably works fine too, without `setenv` -- though I'm trying to figure out why different 🤔 )

The rough part here is when things `chdir` you want an absolute path...

blueyed commented:
Yay! https://codecov.io/gh/pytest-dev/pytest/tree/5abd519e67d7ad889bdbe44fa746af8f3eacac09

Please review the PR via https://github.com/pytest-dev/pytest/pull/2800/files, and I will clean it up after #3877 has been merged.

nicoddemus commented on .travis.yml:
```diff
@@ -62,10 +61,35 @@ jobs:
+      coverage report -m
+      codecov --required -X gcov pycov search -f coverage.xml --flags ${TOXENV//-/ }
+
+      # Coveralls does not support merged reports.
```
I believe we can drop coveralls then. Or would you rather do it in another PR @blueyed?

blueyed replied:
I've thought to keep it for comparison.

nicoddemus replied:
Oh OK. But I think we should remove it after a while to avoid confusion. 👍

nicoddemus commented on appveyor.yml:
```diff
@@ -1,53 +0,0 @@
-environment:
```
I assume you removed AppVeyor so the builds are faster, please remember to bring it back before merging.

nicoddemus commented on tox.ini:
```diff
@@ -17,13 +17,21 @@ envlist =
+    {env:_PYTEST_TOX_COVERAGE_RUN:} pytest --lsof -ra {env:_PYTEST_TEST_OPTS:} {posargs:testing}
+    coverage: coverage report -m --skip-covered
+setenv =
+    coverage: _PYTEST_TOX_COVERAGE_RUN=coverage run -a -m
```
Why do we handle `_PYTEST_TOX_COVERAGE_RUN` and  `_PYTEST_TOX_EXTRA_DEP` both here and in the travis file?

blueyed replied:
This is meant to allow for `tox -e py36-coverage`.
I've thought about/tried using only `-coverage`, but it is not trivial/feasible due to the many different testenvs.

nicoddemus replied:
I see, OK, thanks

blueyed commented:
Squashed (old HEAD: https://github.com/pytest-dev/pytest/pull/2800/commits/78d2c4deb6e6a44236968faf3c31c88e747b3fc9).
I think this is good from my side for now.

blueyed commented:
Meh.. AppVeyor will fail - still using TOXENV=coveralls..

I guess the same mechanism should be used with AppVeyor then?  Basically the `before_script` and `after_success` from Travis would need to be ported then.
@nicoddemus can you do it?  It should also add an extra flag with codecov then, "win" or "windows".

nicoddemus commented:
@blueyed sure I can give it a try. For now I suggest to comment out the `TOXENV=coveralls` in AppVeyor so we can get this merged.

blueyed commented:
Rebased it on master (where the other ci improvements are already), and re-added coveralls for AppVeyor.  Since it also runs codecov there it should be good for Windows for now.

nicoddemus commented:
> Since it also runs codecov there 

You mean `coveralls` right?

blueyed commented:
Both, since 4d19b943.

blueyed commented:
Fixed linting, squashed the keeping of coveralls for AppVeyor:

```diff
diff --git a/.travis.yml b/.travis.yml
index 9279334e..2c9deb3e 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -60,7 +60,7 @@ jobs:
       env: TOXENV=py27
     - env: TOXENV=py34
     - env: TOXENV=py36
-    - env: TOXENV=linting
+    - env: TOXENV=linting PYTEST_NO_COVERAGE=1

     - stage: deploy
       python: '3.6'
```

blueyed commented:
No coverage for doctesting, but works locally (https://travis-ci.org/pytest-dev/pytest/jobs/422108111#L625).
Trying to set `COVERAGE_FILE` for it.

blueyed commented:
doctesting is working now (https://travis-ci.org/pytest-dev/pytest/jobs/422263747#L677), but I've changed the tox testenv quite a bit.

nicoddemus commented:
Thanks a ton @blueyed!

blueyed commented:
Yay! :)
